File: mixed/react-auto-A.java
package net.cell_lang;

import java.util.Arrays;


class ReactAutoA_Tests {
  public static void run() {
    ReactAutoA testAuto = new ReactAutoA();

    boolean aBool = true;
    String aSymb = "my_symbol";
    long anInt = 42;
    double aFloat = 3.14159;
    String aString = "Imagine there's no heaven...";
    long aTimeSpan = 60000;

    boolean[] aBoolSeq = new boolean[] {true, false, false, true};
    long[] anIntSeq = new long[] {100, 101, 102, 103, 104, 105};
    double[] aFloatSeq = new double[] {-2.0, -1.0, 0.0, 1.0, 2.0, 3.0};
    long[] aTimeSpanSeq = new long[] {60, 3600, 86420};
    // String anIntToSymbMap = "[0 -> zero, 1 -> one, 2 -> two, 3 -> three]";
    String anIntToSymbMap = "[1 -> one, 2 -> two, 0 -> zero, 3 -> three]";

    Rect aRect = new Rect();
    aRect.color = Blue.singleton;
    aRect.width = 4.16;
    aRect.height = 0.81;
    aRect.bottomLeft = new Point();
    aRect.bottomLeft.x = 25;
    aRect.bottomLeft.y = 9;

    DateRec aDateRec = new DateRec();
    aDateRec.day = 15;
    aDateRec.month = 3;
    aDateRec.year = 2007;

    Date aDate = new Date();
    aDate.item1 = 5;
    aDate.item2 = 2;
    aDate.item3 = 2018;

    Point aPoint = new Point();
    aPoint.x = 12;
    aPoint.y = 7;

    Polar aPolar = new Polar();
    aPolar.ro = 6.28318;
    aPolar.theta = 1.73205;
    // anAnyPoint = "polar(ro: 6.28318, theta: 1.73205)";

    Double_String_Long_Seq aTuple = new Double_String_Long_Seq();
    aTuple.item1 = 2.71828;
    aTuple.item2 = "It's easy if you try...";
    aTuple.item3 = new long[] {1, 1, 2, 6, 24};

    Point[] aPointSeq = new Point[3];
    for (int i=0 ; i < aPointSeq.length ; i++)
      aPointSeq[i] = new Point();
    aPointSeq[0].x = 0;
    aPointSeq[0].y = 0;
    aPointSeq[1].x = 1;
    aPointSeq[1].y = 1;
    aPointSeq[2].x = 2;
    aPointSeq[2].y = 4;

    // String aRect = "(color: blue, width: 4.16, height: 0.81, bottom_left: point(x: 25, y: 9))";
    // String aDateRec = "(day: 15, year: 2007, month: 3)";
    // var aDate = new Tuple<long, long, long>(5, 2, 2018);
    // String aPoint = "point(x: 12, y: 7)";
    // String anAnyPoint = "polar(ro: 6.28318, theta: 1.73205)";
    // var aTuple = new Tuple<double, String, long[]>(2.71828, "It's easy if you try...", new long[] {1, 1, 2, 6, 24});
    // String[] aPointSeq = new String[] {"point(x: 0, y: 0)", "point(x: 1, y: 1)", "point(x: 2, y: 4)"};

    testAuto.setABoolInput(aBool);
    testAuto.setASymbInput(aSymb);
    testAuto.setAnIntInput(anInt);
    testAuto.setAFloatInput(aFloat);
    testAuto.setAStringInput(aString);
    testAuto.setATimeSpanInput(aTimeSpan);
    testAuto.setARectInput(aRect);
    testAuto.setADateRecInput(aDateRec);
    testAuto.setADateInput(aDate);
    testAuto.setAPointInput(aPoint);
    testAuto.setAnAnyPointInput(aPolar);
    testAuto.setATupleInput(aTuple);
    testAuto.setABoolSeqInput(aBoolSeq);
    testAuto.setAnIntSeqInput(anIntSeq);
    testAuto.setAFloatSeqInput(aFloatSeq);
    testAuto.setAPointSeqInput(aPointSeq);
    testAuto.setATimeSpanSeqInput(aTimeSpanSeq);
    testAuto.setAnIntToSymbMapInput(anIntToSymbMap);

    testAuto.apply();


    boolean aBoolOutput = testAuto.aBoolOutput();
    if (aBoolOutput != aBool) {
      System.out.printf("ERROR: ABoolOutput = %s, ABool = %s\n", Boolean.toString(aBoolOutput), Boolean.toString(aBool));
      return;
    }

    Value aSymbOutput = testAuto.aSymbOutput();
    if (!aSymbOutput.toString().equals(aSymb)) {
      System.out.printf("ERROR: ASymbOutput = %s, ASymb = %s\n", aSymbOutput.toString(), aSymb);
      return;
    }

    long anIntOutput = testAuto.anIntOutput();
    if (anIntOutput != anInt) {
      System.out.printf("ERROR: AnIntOutput = %d, AnInt = %d\n", anIntOutput, anInt);
      return;
    }

    double aFloatOutput = testAuto.aFloatOutput();
    if (aFloatOutput != aFloat) {
      System.out.printf("ERROR: AFloatOutput = %f, AFloat = %f\n", aFloatOutput, aFloat);
      return;
    }

    String aStringOutput = testAuto.aStringOutput();
    if (!aStringOutput.equals(aString)) {
      System.out.printf("ERROR: AStringOutput = %s, AString = %s\n", aStringOutput, aString);
      return;
    }

    long aTimeSpanOutput = testAuto.aTimeSpanOutput();
    if (aTimeSpanOutput != aTimeSpan) {
      System.out.printf("ERROR: ATimeSpanOutput = %d, ATimeSpan = %d\n", aTimeSpanOutput, aTimeSpan);
      return;
    }

    Rect aRectOutput = testAuto.aRectOutput();
    if (!aRectOutput.toString().equals(aRect.toString())) {
      System.out.printf("ERROR: ARectOutput = %s, aRect = %s\n", aRectOutput.toString(), aRect.toString());
      return;
    }

    DateRec aDateRecOutput = testAuto.aDateRecOutput();
    if (!aDateRecOutput.toString().equals(aDateRec.toString())) {
      System.out.printf("ERROR: aDateRecOutput = %s, aDateRec = %s\n", aDateRecOutput.toString(), aDateRec.toString());
      return;
    }

    Date aDateOutput = testAuto.aDateOutput();
    if (aDateOutput.item1 != aDate.item1 || aDateOutput.item2 != aDate.item2 || aDateOutput.item3 != aDate.item3) {
      System.out.printf("ERROR: aDateOutput = %s, aDate = %s\n", aDateOutput.toString(), aDate.toString());
      return;
    }

    Point aPointOutput = testAuto.aPointOutput();
    if (!aPointOutput.toString().equals(aPoint.toString())) {
      System.out.printf("ERROR: aPointOutput = %s, aPoint = %s\n", aPointOutput.toString(), aPoint.toString());
      return;
    }

    AnyPoint anAnyPointOutput = testAuto.anAnyPointOutput();
    if (!anAnyPointOutput.toString().equals(aPolar.toString())) {
      System.out.printf("ERROR: anAnyPointOutput = %s, anAnyPoint = %s\n", anAnyPointOutput.toString(), aPolar.toString());
      return;
    }

    Double_String_Long_Seq aTupleOutput = testAuto.aTupleOutput();
    if (!aTupleOutput.toString().equals(aTuple.toString())) {
      System.out.printf("ERROR: aTupleOutput = %s, aTuple = %s\n", aTupleOutput.toString(), aTuple.toString());
      return;
    }

    boolean[] aBoolSeqOutput = testAuto.aBoolSeqOutput();
    if (!Arrays.equals(aBoolSeqOutput, aBoolSeq)) {
      System.out.printf("ERROR: aBoolSeqOutput = %s, aBoolSeq = %s\n", aBoolSeqOutput.toString(), aBoolSeq.toString());
      return;
    }

    long[] anIntSeqOutput = testAuto.anIntSeqOutput();
    if (!Arrays.equals(anIntSeqOutput, anIntSeq)) {
      System.out.printf("ERROR: anIntSeqOutput = %s, anIntSeq = %s\n", anIntSeqOutput.toString(), anIntSeq.toString());
      return;
    }

    double[] aFloatSeqOutput = testAuto.aFloatSeqOutput();
    if (!Arrays.equals(aFloatSeqOutput, aFloatSeq)) {
      System.out.printf("ERROR: aFloatSeqOutput = %s, aFloatSeq = %s\n", aFloatSeqOutput.toString(), aFloatSeq.toString());
      return;
    }

    Point[] aPointSeqOutput = testAuto.aPointSeqOutput();
    if (!cellEquals(aPointSeqOutput, aPointSeq)) {
      // System.out.printf("ERROR: aPointSeqOutput = %s, aPointSeq = %s\n", aPointSeqOutput.toString(), aPointSeq.toString());
      System.out.println("ERROR: aPointSeqOutput != aPointSeq");

      System.out.println("aPointSeqOutput = [");
      for (int i=0 ; i < aPointSeqOutput.length ; i++)
        System.out.printf("  %s\n", aPointSeqOutput[i].toString());
      System.out.println("]");

      System.out.println("aPointSeq = [");
      for (int i=0 ; i < aPointSeq.length ; i++)
        System.out.printf("  %s\n", aPointSeq[i].toString());
      System.out.println("]");
      return;
    }

    long[] aTimeSpanSeqOutput = testAuto.aTimeSpanSeqOutput();
    if (!Arrays.equals(aTimeSpanSeqOutput, aTimeSpanSeq)) {
      System.out.printf("ERROR: aTimeSpanSeqOutput = %s, aTimeSpanSeq = %s\n", aTimeSpanSeqOutput.toString(), aTimeSpanSeq.toString());
      return;
    }

    Value anIntToSymbMapOutput = testAuto.anIntToSymbMapOutput();
    if (!anIntToSymbMapOutput.toString().equals(anIntToSymbMap)) {
      System.out.printf("ERROR: anIntToSymbMapOutput = %s, anIntToSymbMap = %s\n", anIntToSymbMapOutput.toString(), anIntToSymbMap);
      return;
    }

    System.out.println("Reactive automata tests passed!");
  }

  static boolean cellEquals(Object[] xs, Object[] ys) {
    if (xs.length != ys.length)
      return false;
    for (int i=0 ; i < xs.length ; i++)
      if (!xs[i].toString().equals(ys[i].toString()))
        return false;
    return true;
  }
}

File: mixed/react-auto-A.cs
using System;
using System.IO;

using Cell.Facades;


public static class ReactAutoA_Tests {
  public static void Run() {
    ReactAutoA testAuto = new ReactAutoA();

    bool aBool = true;
    string aSymb = "my_symbol";
    long anInt = 42;
    double aFloat = 3.14159;
    string aString = "Imagine there's no heaven...";
    long aTimeSpan = 60000;
    var aRect = (color: Blue.singleton, width: 4.16, height: 0.81, bottomLeft: (x: 25, y: 9));
    // string aRect = "(color: blue, width: 4.16, height: 0.81, bottom_left: point(x: 25, y: 9))";
    var aDateRec = (day: 15, year: 2007, month: 3);
    // string aDateRec = "(day: 15, year: 2007, month: 3)";
    (int, int, int) aDate = (5, 2, 2018);

    var aPoint = (x: 12, y: 7);
    // string aPoint = "point(x: 12, y: 7)";

    Polar polar = new Polar();
    polar.ro = 6.28318;
    polar.theta = 1.73205;
    AnyPoint anAnyPoint = polar;
    // string anAnyPoint = "polar(ro: 6.28318, theta: 1.73205)";

    var aTuple = (2.71828, "It's easy if you try...", new long[] {1, 1, 2, 6, 24});
    bool[] aBoolSeq = new bool[] {true, false, false, true};
    long[] anIntSeq = new long[] {100, 101, 102, 103, 104, 105};
    double[] aFloatSeq = new double[] {-2.0, -1.0, 0.0, 1.0, 2.0, 3.0};
    string[] aPointSeq = new string[] {"point(x: 0, y: 0)", "point(x: 1, y: 1)", "point(x: 2, y: 4)"};
    string _aPointSeq = "(point(x: 0, y: 0), point(x: 1, y: 1), point(x: 2, y: 4))";
    long[] aTimeSpanSeq = new long[] {60, 3600, 86420};
    // string anIntToSymbMap = "[0 -> zero, 1 -> one, 2 -> two, 3 -> three]";
    string anIntToSymbMap = "[1 -> one, 2 -> two, 0 -> zero, 3 -> three]";

    testAuto.ABoolInput = aBool;
    testAuto.ASymbInput = aSymb;
    testAuto.AnIntInput = anInt;
    testAuto.AFloatInput = aFloat;
    testAuto.AStringInput = aString;
    testAuto.ATimeSpanInput = aTimeSpan;
    testAuto.ARectInput = aRect;
    // testAuto.SetInput(ReactAutoA.Input.A_RECT_INPUT, aRect);
    testAuto.ADateRecInput = aDateRec;
    // testAuto.SetInput(ReactAutoA.Input.A_DATE_REC_INPUT, aDateRec);
    testAuto.ADateInput = aDate;
    // testAuto.SetInput(ReactAutoA.Input.A_DATE_INPUT, _aDate);
    testAuto.APointInput = aPoint;
    // testAuto.SetInput(ReactAutoA.Input.A_POINT_INPUT, aPoint);
    testAuto.AnAnyPointInput = anAnyPoint;
    // testAuto.SetInput(ReactAutoA.Input.AN_ANY_POINT_INPUT, anAnyPoint);
    testAuto.ATupleInput = aTuple;
    // testAuto.SetInput(ReactAutoA.Input.A_TUPLE_INPUT, _aTuple);
    testAuto.ABoolSeqInput = aBoolSeq;
    testAuto.AnIntSeqInput = anIntSeq;
    testAuto.AFloatSeqInput = aFloatSeq;
    // testAuto.APointSeqInput = aPointSeq;
    testAuto.SetInput(ReactAutoA.Input.A_POINT_SEQ_INPUT, _aPointSeq);
    testAuto.ATimeSpanSeqInput = aTimeSpanSeq;
    // testAuto.AnIntToSymbMapInput = anIntToSymbMap;
    testAuto.SetInput(ReactAutoA.Input.AN_INT_TO_SYMB_MAP_INPUT, anIntToSymbMap);

    testAuto.Apply();

    if (testAuto.ABoolOutput != aBool) {
      Console.WriteLine("ERROR: ABoolOutput = {0}, ABool = {1}", testAuto.ABoolOutput, aBool);
      return;
    }

    if (testAuto.ASymbOutput != aSymb) {
      Console.WriteLine("ERROR: ASymbOutput = {0}, ASymb = {1}", testAuto.ASymbOutput, aSymb);
      return;
    }

    if (testAuto.AnIntOutput != anInt) {
      Console.WriteLine("ERROR: AnIntOutput = {0}, AnInt = {1}", testAuto.AnIntOutput, anInt);
      return;
    }

    if (testAuto.AFloatOutput != aFloat) {
      Console.WriteLine("ERROR: AFloatOutput = {0}, AFloat = {1}", testAuto.AFloatOutput, aFloat);
      return;
    }

    if (testAuto.AStringOutput != aString) {
      Console.WriteLine("ERROR: AStringOutput = {0}, AString = {1}", testAuto.AStringOutput, aString);
      return;
    }

    if (testAuto.ATimeSpanOutput != aTimeSpan) {
      Console.WriteLine("ERROR: ATimeSpanOutput = {0}, ATimeSpan = {1}", testAuto.ATimeSpanOutput, aTimeSpan);
      return;
    }

    var outRect = testAuto.ARectOutput;
    if (outRect != aRect) {
      Console.WriteLine("ERROR: ARectOutput = {0}, aRect = {1}", outRect, aRect);
      return;
    }

    var outDateRec = testAuto.ADateRecOutput;
    if (outDateRec != aDateRec) {
      Console.WriteLine("ERROR: aDateRecOutput = {0}, aDateRec = {1}", outDateRec, aDateRec);
      return;
    }

    if (testAuto.ADateOutput.Item1 != aDate.Item1) {
      Console.WriteLine("ERROR: ADateOutput.Item1 = {0}, ADate.Item1 = {1}", testAuto.ADateOutput.Item1, aDate.Item1);
      return;
    }

    if (testAuto.ADateOutput.Item2 != aDate.Item2) {
      Console.WriteLine("ERROR: ADateOutput.Item2 = {0}, ADate.Item2 = {1}", testAuto.ADateOutput.Item2, aDate.Item2);
      return;
    }

    if (testAuto.ADateOutput.Item3 != aDate.Item3) {
      Console.WriteLine("ERROR: ADateOutput.Item3 = {0}, ADate.Item3 = {1}", testAuto.ADateOutput.Item3, aDate.Item3);
      return;
    }

    var outPoint = testAuto.APointOutput;
    if (outPoint != aPoint) {
      Console.WriteLine("ERROR: APointOutput = {0}, APoint = {1}", outPoint, aPoint);
      return;
    }

    if (testAuto.AnAnyPointOutput.ToString() != anAnyPoint.ToString()) {
      Console.WriteLine("ERROR: AnAnyPointOutput = {0}, AnAnyPoint = {1}", testAuto.AnAnyPointOutput, anAnyPoint);
      return;
    }

    if (testAuto.ATupleOutput.Item1 != aTuple.Item1) {
      Console.WriteLine("ERROR: ATupleOutput.Item1 = {0}, ATuple.Item1 = {1}", testAuto.ATupleOutput.Item1, aTuple.Item1);
      return;
    }

    if (testAuto.ATupleOutput.Item2 != aTuple.Item2) {
      Console.WriteLine("ERROR: ATupleOutput.Item2 = {0}, ATuple.Item2 = {1}", testAuto.ATupleOutput.Item2, aTuple.Item2);
      return;
    }

    if (!Eq(testAuto.ATupleOutput.Item3, aTuple.Item3)) {
      Console.WriteLine("ERROR: ATupleOutput.Item3 = {0}, ATuple.Item3 = {1}", testAuto.ATupleOutput.Item3, aTuple.Item3);
      return;
    }

    if (!Eq(testAuto.ABoolSeqOutput, aBoolSeq)) {
      Console.WriteLine("ERROR: ABoolSeqOutput = {0}, ABoolSeq = {1}", testAuto.ABoolSeqOutput, aBoolSeq);
      return;
    }

    if (!Eq(testAuto.AnIntSeqOutput, anIntSeq)) {
      Console.WriteLine("ERROR: AnIntSeqOutput = {0}, AnIntSeq = {1}", testAuto.AnIntSeqOutput, anIntSeq);
      return;
    }

    if (!Eq(testAuto.AFloatSeqOutput, aFloatSeq)) {
      Console.WriteLine("ERROR: AFloatSeqOutput = {0}, AFloatSeq = {1}", testAuto.AFloatSeqOutput, aFloatSeq);
      return;
    }


    var outPointSeq = testAuto.APointSeqOutput;
    string[] strs = new string[outPointSeq.Length];
    for (int i=0 ; i < strs.Length ; i++)
      strs[i] = string.Format("point(x: {0}, y: {1})", outPointSeq[i].x, outPointSeq[i].y);
    if (!EqStrV(strs, aPointSeq)) {
      Console.WriteLine("ERROR: APointSeqOutput = {0}, APointSeq = {1}", strs, _aPointSeq);
      return;
    }

    if (!Eq(testAuto.ATimeSpanSeqOutput, aTimeSpanSeq)) {
      Console.WriteLine("ERROR: ATimeSpanSeqOutput = {0}, ATimeSpanSeq = {1}", testAuto.ATimeSpanSeqOutput, aTimeSpanSeq);
      return;
    }

    string str = "[";
    foreach (var entry in testAuto.AnIntToSymbMapOutput) {
      if (str != "[")
        str += ", ";
      str += string.Format("{0} -> {1}", entry.Key, entry.Value);
    }
    str += "]";
    if (str != anIntToSymbMap) {
      Console.WriteLine("ERROR: AnIntToSymbMapOutput = {0}, anIntToSymbMap = {1}", str, anIntToSymbMap);
      return;
    }

    Console.WriteLine("Reactive automata tests passed!");
  }

  static bool Eq<T>(T[] xs, T[] ys) where T : System.IComparable<T> {
    if (xs.Length != ys.Length)
      return false;
    for (int i=0 ; i < xs.Length ; i++)
      if (xs[i].CompareTo(ys[i]) != 0)
        return false;
    return true;
  }

  static bool EqStrV<T>(T[] xs, string[] ys) {
    if (xs.Length != ys.Length)
      return false;
    for (int i=0 ; i < xs.Length ; i++)
      if (xs[i].ToString() != ys[i])
        return false;
    return true;
  }
}
